peachjam/views/save_document.py

```diff
@@ -0,0 +1,153 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth.mixins import LoginRequiredMixin, PermissionRequiredMixin
+from django.http import Http404, HttpResponse, HttpResponseRedirect
+from django.shortcuts import get_object_or_404
+from django.urls import reverse, reverse_lazy
+from django.views.generic import (
+    CreateView,
+    DeleteView,
+    ListView,
+    TemplateView,
+    UpdateView,
+)
+from django.views.generic.edit import FormMixin
+
+from peachjam.forms import FolderForm, SaveDocumentForm
+from peachjam.models import CoreDocument, Folder, SavedDocument, pj_settings
+
+User = get_user_model()
+
+
+class AllowSavedDocumentMixin:
+    def dispatch(self, *args, **kwargs):
+        if not pj_settings().allow_save_documents:
+
+            raise Http404("Saving documents is not allowed.")
+        return super().dispatch(*args, **kwargs)
+
+
+class BaseFolderView(
+    AllowSavedDocumentMixin, LoginRequiredMixin, PermissionRequiredMixin
+):
+    model = Folder
+
+    def get_template_names(self):
+        if self.request.htmx:
+            return ["peachjam/_folders_list.html"]
+        return ["peachjam/folders_list.html"]
+
+    def get_queryset(self):
+        return self.request.user.folders.all()
+
+    def get_context_data(self, *args, **kwargs):
```

Comment:
Be careful about abusing a mixin. If this logic is only used by one subclass, then it shouldn't be here.

A rule of thumb is to implement all the classes, see what's duplicated, and only if it's used more than two times, then create a mixin. Otherwise it gets complicated to work out when the various mixin methods are applicable.

Reply:
Right, I've moved the get_context_data to the list view where it is used.